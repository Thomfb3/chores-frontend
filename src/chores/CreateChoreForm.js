import React, { useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import AppAlert from "../common/AppAlert"
import UserContext from "../auth/UserContext";
import TextField from '@mui/material/TextField';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import Select from '@mui/material/Select';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import DateTimePicker from '@mui/lab/DateTimePicker';
import Button from '@mui/material/Button';
import Paper from '@mui/material/Paper';


function CreateChoreForm({ createChore }) {
    const history = useHistory();
    const { currentUser, currentTeam, currentTeamUsers } = useContext(UserContext);
    const [dueDateValue, setDueDateValue] = useState(new Date());
    const [formData, setFormData] = useState({
        "title": "",
        "description": "",
        "pointValue": "",
        "assigner": currentUser._id,
        "assignee": "unassigned",
        "createdBy": currentUser._id,
        "dueDate": "",
        "status": "created",
        "imageCover": "default-chore.jpg",
        "teamId": currentTeam._id,
        "type": "template"
    });

    const [formErrors, setFormErrors] = useState([]);
    console.log(formErrors);
    console.debug(
        "CreateChoreForm",
        "createChore=", typeof createChore,
        "formData=", formData,
        "formErrors", formErrors,
    );

    async function handleSubmit(evt) {
        evt.preventDefault();
        formData.imageCover = (formData.imageCover=== "") ? "default-chore.jpg" : formData.imageUrl;
        formData.assignee = formData.assignee === "unassigned" ? null : formData.assignee;
        let result = await createChore(formData);
        if (result.success) {
            history.push("/chores");
        } else {
            setFormErrors(result.errors);
        };
    };

    function handleChange(evt) {
        formData.dueDate = dueDateValue;
        const { name, value } = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    };

    return (
        <div className="Form">
            <div className="Form__title">
                <div className="Form__list-title">Manage Chores</div>
                <div className="Form__divider"></div>
            </div>
            <div className="Form__form-container">
                <Paper
                    elevation={3}
                    sx={{
                        width: '50%', paddingTop: '30px', margin: 'auto', marginTop: '50px', paddingBottom: '10px',
                        '@media (max-width:900px)': {
                            width: '90%'
                        }
                    }}>
                    <form className="Form" onSubmit={handleSubmit}>
                        <div className="Form__box-title">Create New Chore</div>
                        <div className="Form-group">
                            <TextField
                                id="title"
                                label="Title"
                                variant="outlined"
                                placeholder="  "
                                type="text"
                                name="title"
                                value={formData.title}
                                onChange={handleChange}
                                sx={{ m: 1, width: '95%' }}
                                required />
                        </div>
                        <div className="Form-group">
                            <TextField
                                id="description"
                                label="Description"
                                variant="outlined"
                                placeholder="  "
                                type="text"
                                name="description"
                                value={formData.description}
                                onChange={handleChange}
                                sx={{ m: 1, width: '95%' }}
                            />
                        </div>
                        <div className="Form-group">
                            <TextField
                                id="pointValue"
                                label="Point Value"
                                variant="outlined"
                                placeholder="  "
                                type="number"
                                name="pointValue"
                                value={formData.pointValue}
                                onChange={handleChange}
                                sx={{ m: 1, width: '95%' }}
                            />
                        </div>
                        <div className="Form-group">
                            <TextField
                                id="imageCover"
                                label="Chore Image"
                                variant="outlined"
                                placeholder="  "
                                type="text"
                                name="imageCover"
                                value={formData.imageCover}
                                onChange={handleChange}
                                sx={{ m: 1, width: '95%' }}
                                />
                        </div>

                        <FormControl sx={{ m: 1, width: '95%' }}>
                            <InputLabel id="assignee" sx={{ backgroundColor: 'white', padding: '3px' }}>Assign To</InputLabel>
                            <Select
                                labelId="assignee"
                                id="assignee-select"
                                value={formData.assignee}
                                name="assignee"
                                onChange={handleChange}
                            >
                                <MenuItem key={"unassigned"} value={"unassigned"}>Unassigned</MenuItem>
                                {currentTeamUsers.map(u => (
                                    <MenuItem key={u._id} value={u._id}>{u.username}</MenuItem>

                                ))}
                            </Select>
                        </FormControl>

                        <LocalizationProvider name="dueDate" dateAdapter={AdapterDateFns}>
                            <DateTimePicker
                                renderInput={(props) => <TextField {...props} sx={{ m: 1, width: '95%' }} />}
                                type="datetime-local"
                                name="dueDate"
                                label="Due Date"
                                value={dueDateValue}
                                onChange={(newDueDateValue) => {
                                    setDueDateValue(newDueDateValue);
                                }}
                            />
                        </LocalizationProvider>

                        {formErrors.length
                            ? <AppAlert severity="error" messages={formErrors} />
                            : null}

                        <div className="Form-group">
                            <Button
                                sx={{ m: 1, backgroundColor: '#1193ff', borderRadius: '5px' }}
                                variant="contained"
                                type="submit"
                                onSubmit={handleSubmit}
                            >
                                Create Chore
                            </Button>
                        </div>
                    </form>
                </Paper>
            </div>
        </div>
    );
};

export default CreateChoreForm;