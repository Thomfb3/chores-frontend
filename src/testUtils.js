import React from "react";
import UserContext from "./auth/UserContext";
import ChoreCommentContext from "./comments/ChoreCommentContext";

const demoUser = {
  _id: "userid1",
  username: "testuser1",
  firstName: "testfirst1",
  lastName: "testlast1",
  profileImage: "defaultProfile.jpg",
  allTimePoints: 1000,
  currentPoints: 200,
  role: "admin"
};

const testTeam = {
  _id: "testTeam",
  name: "testTeam"
};

const testTeamUsers = [
  {
    _id: "userid1",
    username: "testuser1",
    firstName: "testfirst1",
    lastName: "testlast1",
    profileImage: "defaultProfile.jpg",
    allTimePoints: 1000,
    currentPoints: 200,
    role: "admin"
  },
  {
    _id: "userid2",
    username: "testuser2",
    firstName: "testfirst2",
    lastName: "testlast2",
    profileImage: "defaultProfile.jpg",
    allTimePoints: 1000,
    currentPoints: 200,
    role: "user"
  },
  {
    _id: "userid3",
    username: "testuser3",
    firstName: "testfirst3",
    lastName: "testlast3",
    profileImage: "defaultProfile.jpg",
    allTimePoints: 1000,
    currentPoints: 200,
    role: "user"
  },
  {
    _id: "userid4",
    username: "testuser4",
    firstName: "testfirst4",
    lastName: "testlast4",
    profileImage: "defaultProfile.jpg",
    allTimePoints: 1000,
    currentPoints: 200,
    role: "user"
  }
];

let choreComments = [];

const UserProvider =
  ({ children,
    currentUser = demoUser,
    currentTeam = testTeam,
    currentTeamUsers = testTeamUsers,
    isAdmin = false
  }) => (
      <UserContext.Provider value={{ currentUser, currentTeam, currentTeamUsers }}>
        {children}
      </UserContext.Provider>
    );


const CommentProvider =
  ({ children,
    currentChoreComments = choreComments
  }) => (
      <ChoreCommentContext.Provider value={{ currentChoreComments }}>
        {children}
      </ChoreCommentContext.Provider>
    );

export { UserProvider, CommentProvider };

