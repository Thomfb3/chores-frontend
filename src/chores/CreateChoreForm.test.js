import React from "react";
import { render } from "@testing-library/react";
import CreateChoreForm from "./CreateChoreForm";
import { UserProvider } from "../testUtils";


it("renders without crashing", function () {
    render(
          <UserProvider>
            <CreateChoreForm />
          </UserProvider>
    );
  });
