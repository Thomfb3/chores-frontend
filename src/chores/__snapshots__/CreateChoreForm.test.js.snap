// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Form"
  >
    <div
      class="Form__title"
    >
      <div
        class="Form__list-title"
      >
        Manage Chores
      </div>
      <div
        class="Form__divider"
      />
    </div>
    <div
      class="Form__form-container"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-5xxjkm-MuiPaper-root"
      >
        <form
          class="Form"
        >
          <div
            class="Form__box-title"
          >
            Create New Chore
          </div>
          <div
            class="Form-group"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-nkaxf5-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="title"
                id="title-label"
              >
                Title
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="title"
                  name="title"
                  placeholder="  "
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Title *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="Form-group"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-nkaxf5-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="description"
                id="description-label"
              >
                Description
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="description"
                  name="description"
                  placeholder="  "
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Description
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="Form-group"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-nkaxf5-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="pointValue"
                id="pointValue-label"
              >
                Point Value
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="pointValue"
                  name="pointValue"
                  placeholder="  "
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Point Value
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="Form-group"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-nkaxf5-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="imageCover"
                id="imageCover-label"
              >
                Chore Image
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="imageCover"
                  name="imageCover"
                  placeholder="  "
                  type="text"
                  value="default-chore.jpg"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1z7n62"
                  >
                    <span>
                      Chore Image
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiFormControl-root css-1p2pgsw-MuiFormControl-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-18gb8ry-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="assignee"
            >
              Assign To
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="assignee assignee-select"
                class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="assignee-select"
                role="button"
                tabindex="0"
              >
                Unassigned
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                name="assignee"
                tabindex="-1"
                value="unassigned"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-nnbavb"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-nkaxf5-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
            >
              Due Date
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                aria-label="Choose date, selected date is Dec 9, 2021"
                aria-readonly="true"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                type="text"
                value="12/09/2021 11:59 am"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1z7n62"
                >
                  <span>
                    Due Date
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="Form-group"
            style="text-align: right;"
          >
            <button
              class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-58dpzn-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              tabindex="0"
              type="submit"
            >
              Create Chore
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
